Use billing cycles fetch subscriptions month

ashwin003 opened the pull request:
Closes #87

ashwin003 committed:
Saves billing occurrence and intervals while creating a new subscription

ashwin003 committed:
Saves billing occurrence and intervals while updating subscriptions

ashwin003 committed:
Uses Billing Cycles while fetching subscriptions for a month

ashwin003 commented:
@chrisK00 not sure which test is failing and why. SOS!

chrisK00 commented:
> @chrisK00 not sure which test is failing and why. SOS!

One of the autopaid tests. Says stackoverflow. Did you try running the tests locally?

ashwin003 commented:
> > @chrisK00 not sure which test is failing and why. SOS!
> 
> One of the autopaid tests. Says stackoverflow. Did you try running the tests locally?

Just the ones I updated/added. Nothing was changed on the autopay side. Why is it failing?

ashwin003 committed:
Updates subscription billing cycle values for test data

chrisK00 commented:
> > > @chrisK00 not sure which test is failing and why. SOS!
> > 
> > One of the autopaid tests. Says stackoverflow. Did you try running the tests locally?
> 
> Just the ones I updated/added. Nothing was changed on the autopay side. Why is it failing?

Try running all tests.  Did you check what methods call the code you changed?

ashwin003 commented:
> > > > @chrisK00 not sure which test is failing and why. SOS!
> > > 
> > > 
> > > One of the autopaid tests. Says stackoverflow. Did you try running the tests locally?
> > 
> > 
> > Just the ones I updated/added. Nothing was changed on the autopay side. Why is it failing?
> 
> Try running all tests. Did you check what methods call the code you changed?

Glad that cleared itself.

chrisK00 commented:
> > > > > @chrisK00 not sure which test is failing and why. SOS!
> > > > 
> > > > 
> > > > One of the autopaid tests. Says stackoverflow. Did you try running the tests locally?
> > > 
> > > 
> > > Just the ones I updated/added. Nothing was changed on the autopay side. Why is it failing?
> > 
> > Try running all tests. Did you check what methods call the code you changed?
> 
> Glad that cleared itself.

Nice lol

chrisK00 commented on subtrack.MAUI/Services/SubscriptionsCalculator.cs:
```diff
@@ -95,4 +91,31 @@ public DateTime GetNextPaymentDate(Subscription subscription)
+
+    private static bool PaidThisMonth(Subscription subscription, DateTime monthDate) => subscription.LastPayment.Month == monthDate.Month && subscription.LastPayment.Year == monthDate.Year;
+
+    private static bool PaidInPast(Subscription subscription, DateTime monthDate) => (subscription.LastPayment.Month < monthDate.Month && subscription.LastPayment.Year == monthDate.Year) || subscription.LastPayment.Year < monthDate.Year;
```
This doesn't work for a weekly subscription that has last payment this month. If im not mistaken we need access to the next payment in order to know if it should be paid this month

ashwin003 replied:
It won't fetch any subscriptions that have last payment this month. This is based on the existing behaviour.

```
 var subscriptionsByMonth = subscriptions
                    .Where(s => !paidThisMonth(s) && !paymentStartsInTheFuture(s))
                    .ToList();
 ```

chrisK00 replied:
Yes because we only had monthly subs before so if a sub had been paid this month wed know that it shouldnt be included. We need to add a special case for weekly subs

chrisK00 commented on subtrack.MAUI/Services/SubscriptionsCalculator.cs:
```diff
@@ -95,4 +91,31 @@ public DateTime GetNextPaymentDate(Subscription subscription)
+    {
+        var subscriptions = new List<Subscription>();
+        var subscriptionToUse = (Subscription)subscription.Clone();
+        while (PaidInPast(GetNextSubscription(subscriptionToUse), monthDate))
```
Dont we already know its paid in past since we filtered away thoose that are not?

ashwin003 replied:
we do. we're bringing that past subscription to the present with this loop. 
If a weekly subscription started in the first week of July (let's say the last payment is 7 Jul) and we're looking at it sometime around mid-August (14 Aug), we don't need the instances where next payment is in July while displaying subscriptions for August.

chrisK00 commented on subtrack.MAUI/Services/SubscriptionsCalculator.cs:
```diff
@@ -95,4 +91,31 @@ public DateTime GetNextPaymentDate(Subscription subscription)
+
+    private static bool PaidThisMonth(Subscription subscription, DateTime monthDate) => subscription.LastPayment.Month == monthDate.Month && subscription.LastPayment.Year == monthDate.Year;
+
+    private static bool PaidInPast(Subscription subscription, DateTime monthDate) => (subscription.LastPayment.Month < monthDate.Month && subscription.LastPayment.Year == monthDate.Year) || subscription.LastPayment.Year < monthDate.Year;
```
weekly sub, last payment august 19 - next should be august 26 doesnt show up. Need to consider the occurrence of the sub. Probs need to change PaidThisMonth too

chrisK00 commented on subtrack.MAUI/Services/SubscriptionsCalculator.cs:
```diff
@@ -95,4 +91,31 @@ public DateTime GetNextPaymentDate(Subscription subscription)
+        while (PaidThisMonth(GetNextSubscription(subscriptionToUse), monthDate))
+        {
+            subscriptions.Add((Subscription)subscriptionToUse.Clone());
+            subscriptionToUse.LastPayment = GetNextPaymentDate(subscriptionToUse);
```
this is called inside GetNextSub already

chrisK00 commented on subtrack.MAUI/Services/SubscriptionsCalculator.cs:
```diff
@@ -51,14 +51,10 @@ public decimal GetYearlyCost(Subscription subscription)
+        return subscriptions
+                    .Where(subscription => PaidInPast(subscription, monthDate))
+                    .SelectMany(subscription => GetIterationsForMonth(subscription, monthDate))
+                    .OrderBy(subscription => subscription.LastPayment);
```
Head in to the Monthly.razor and remove the unneeded order by there.  Linq is beautiful

chrisK00 commented on subtrack.Tests/SubscriptionCalculatorTests/GetSubscriptionListByMonthTests.cs:
```diff
@@ -50,25 +53,57 @@ public void GetSubscriptionListByMonth_MonthProvided_ShouldNotReturn_PaidSubscri
+            var result = _sut.GetSubscriptionListByMonth(subscriptions, today.AddMonths(2)).ToList();
+
+            // Assert
+            Assert.True(result.Count(s => s.Name == subscriptions[5].Name) == 4);
```
place the magic numbers inside variables, can be difficult to figure out a good name but since this is a test the name can be how long you want

chrisK00 commented on subtrack.Tests/SubscriptionCalculatorTests/GetSubscriptionListByMonthTests.cs:
```diff
@@ -50,25 +53,57 @@ public void GetSubscriptionListByMonth_MonthProvided_ShouldNotReturn_PaidSubscri
+            IList<Subscription> subscriptions = CreateSubscriptions();
+
+            // Act
+            List<Subscription> first = _sut.GetSubscriptionListByMonth(subscriptions, today.AddMonths(2)).ToList(),
```
This feels like it should either be multiple tests or a theory

ashwin003 committed:
Refactoring, and weekly subscriptions update

chrisK00 commented on subtrack.MAUI/Services/SubscriptionsCalculator.cs:
```diff
@@ -51,14 +51,10 @@ public decimal GetYearlyCost(Subscription subscription)
-
-        return subscriptionsByMonth;
+        return subscriptions
+                    .Where(subscription => PaidInPast(subscription, monthDate) || (subscription.BillingOccurrence == BillingOccurrence.Week && PaidThisMonth(subscription, monthDate)))
```
This logic is a bit broken, the previously mentioned sub now displays its last payment on the monthly page. Im gonna try to see how it can all be written in a more readable manner

chrisK00 committed:
add test case for week and use global usings

chrisK00 committed:
rename

chrisK00 committed:
remove mapping code in monthly.razor. simplify get subscription list.

chrisK00 commented:
@ashwin003 Updated the logic for the feature so you just need to modify the tests, feel free to review the added code aswell. If some tests fail lmk in case there is a bug hiding in the logic.

The reasoning behind returning the monthly lists directly is that we dont have to for every month, calculate the same next payment dates for the same subscription multiple times. Instead we just calculate all payment dates between a given interval and then group all payments that are in the same month. I also refactored out the code from the view to the service so the view can simply be a view. Before changing anything I also added a test case for the weekly sub Ive previously mentioned

chrisK00 committed:
date instead of monthdate

ashwin003 commented:
> @ashwin003 Updated the logic for the feature so you just need to modify the tests, feel free to review the added code aswell. If some tests fail lmk in case there is a bug hiding in the logic.
> 
> The reasoning behind returning the monthly lists directly is that we dont have to for every month, calculate the same next payment dates for the same subscription multiple times. Instead we just calculate all payment dates between a given interval and then group all payments that are in the same month. I also refactored out the code from the view to the service so the view can simply be a view. Before changing anything I also added a test case for the weekly sub Ive previously mentioned

At this point, the build itself is failing as the tests still expect `ISubscriptionsCalculator` to contain the old `GetSubscriptionListByMonth` method

chrisK00 commented:
> > @ashwin003 Updated the logic for the feature so you just need to modify the tests, feel free to review the added code aswell. If some tests fail lmk in case there is a bug hiding in the logic.
> > 
> > The reasoning behind returning the monthly lists directly is that we dont have to for every month, calculate the same next payment dates for the same subscription multiple times. Instead we just calculate all payment dates between a given interval and then group all payments that are in the same month. I also refactored out the code from the view to the service so the view can simply be a view. Before changing anything I also added a test case for the weekly sub Ive previously mentioned
> 
> At this point, the build itself is failing as the tests still expect `ISubscriptionsCalculator` to contain the old `GetSubscriptionListByMonth` method

Its not failing for me, github action ran fine aswell. I could run the app no probs. Try a rebuild? Or just update the tests without doing a build // comment out tests then update one at a time

ashwin003 commented:
> > > @ashwin003 Updated the logic for the feature so you just need to modify the tests, feel free to review the added code aswell. If some tests fail lmk in case there is a bug hiding in the logic.
> > > The reasoning behind returning the monthly lists directly is that we dont have to for every month, calculate the same next payment dates for the same subscription multiple times. Instead we just calculate all payment dates between a given interval and then group all payments that are in the same month. I also refactored out the code from the view to the service so the view can simply be a view. Before changing anything I also added a test case for the weekly sub Ive previously mentioned
> > 
> > 
> > At this point, the build itself is failing as the tests still expect `ISubscriptionsCalculator` to contain the old `GetSubscriptionListByMonth` method
> 
> Its not failing for me, github action ran fine aswell. I could run the app no probs. Try a rebuild? Or just update the tests without doing a build // comment out tests then update one at a time

![image](https://github.com/Code2Gether-Discord/subtrack/assets/12110372/cfbb6633-9629-49b7-a29d-1ddfb8d5318c)


![image](https://github.com/Code2Gether-Discord/subtrack/assets/12110372/1ff72260-82ac-4dae-91dd-9f7ddad07cbb)

chrisK00 commented:
> > > > @ashwin003 Updated the logic for the feature so you just need to modify the tests, feel free to review the added code aswell. If some tests fail lmk in case there is a bug hiding in the logic.
> > > > The reasoning behind returning the monthly lists directly is that we dont have to for every month, calculate the same next payment dates for the same subscription multiple times. Instead we just calculate all payment dates between a given interval and then group all payments that are in the same month. I also refactored out the code from the view to the service so the view can simply be a view. Before changing anything I also added a test case for the weekly sub Ive previously mentioned
> > > 
> > > 
> > > At this point, the build itself is failing as the tests still expect `ISubscriptionsCalculator` to contain the old `GetSubscriptionListByMonth` method
> > 
> > Its not failing for me, github action ran fine aswell. I could run the app no probs. Try a rebuild? Or just update the tests without doing a build // comment out tests then update one at a time
> 
> ![image](https://github.com/Code2Gether-Discord/subtrack/assets/12110372/cfbb6633-9629-49b7-a29d-1ddfb8d5318c)
> 
> 
> ![image](https://github.com/Code2Gether-Discord/subtrack/assets/12110372/1ff72260-82ac-4dae-91dd-9f7ddad07cbb)
> 

Thats only the tests failing, the actual app works fine. As i mentioned in my comment youll need to update the tests since I changed the flow of how getting the sub list works

ashwin003 committed:
Minor refactorings

ashwin003 committed:
Updates tests for fetching monthly subscriptions

chrisK00 commented on subtrack.MAUI/Utilities/DateTimeExtensions.cs:
```diff
@@ -1,19 +1,18 @@
+
+    public static string MonthName(this DateTime date) => date.ToString("MMMM");
+
+    public static DateTime LastDayOfMonth(this DateTime date) => new(date.Year, date.Month, DateTime.DaysInMonth(date.Year, date.Month));
```
This is nice, maybe change the name to be LastDayOfMonthDate since otherwise one could expect an int to be returned

ashwin003 committed:
Refactoring: rename date extensions

chrisK00 commented on subtrack.MAUI/Services/SubscriptionsCalculator.cs:
```diff
@@ -49,16 +50,30 @@ public decimal GetYearlyCost(Subscription subscription)
+                  new SubscriptionsMonthResponse
+                  {
+                      MonthDate = new DateTime(g.Key.Year, g.Key.Month, 1),
+                      Subscriptions = g.OrderBy(s => s.LastPayment),
```
Better to do one .ToList here then the caller making 3x, or more because they might not assign the value to a variable menaing it will still be a pure IEnumerable with an order by waiting to go through each item in the list. The result from this should be complete and directly usable by the view.

chrisK00 commented on subtrack.MAUI/Services/SubscriptionsCalculator.cs:
```diff
@@ -49,16 +50,30 @@ public decimal GetYearlyCost(Subscription subscription)
+                      MonthDate = new DateTime(g.Key.Year, g.Key.Month, 1),
+                      Subscriptions = g.OrderBy(s => s.LastPayment),
+                      Cost = GetTotalCost(g)
+                  });
```
Not doing a .ToList() here breaks the view since the caller uses the IEnumerable and then filters away multiple times. Try running the app and heading to monthly, items for 2024 and 2025 wont be shown. Better to do one .ToList() here then have the caller do 3x without them knowing that it will execute some long queries. Consider this method an API response rather than an internal method, the view shouldn't need to know how it works it just gets the data.

chrisK00 commented on subtrack.Tests/Utilities/SubscriptionEqualityComparer.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    internal class SubscriptionEqualityComparer : IEqualityComparer<Subscription>
+    {
+        public bool Equals(Subscription? x, Subscription? y)
```
its fine to just check the amount of times an item is shown in a list to make sure it exists, e.g. by its ID since that is the way to know if 2 items are the same or not. The result is basically the same but without this class so this can probs be removed

ashwin003 committed:
Use Id to compare subscriptions in tests

ashwin003 committed:
Adds ToList in fetching subscriptions

chrisK00 commented:
Click here to continue the streak --->Squat and Merge<---
